README.md
# gridsome-plugin-manifest

Generates a manifest and handle icons generation for your PWA.

[![npm](https://img.shields.io/npm/v/gridsome-plugin-manifest)](https://www.npmjs.com/package/gridsome-plugin-manifest) [![npm peer dependency version](https://img.shields.io/npm/dependency-version/gridsome-plugin-manifest/peer/gridsome)](https://www.npmjs.com/package/gridsome) [![NPM](https://img.shields.io/npm/l/gridsome-plugin-manifest)](https://github.com/khalyomede/gridsome-plugin-manifest/blob/master/LICENSE) [![Build Status](https://travis-ci.com/khalyomede/gridsome-plugin-manifest.svg?branch=master)](https://travis-ci.com/khalyomede/gridsome-plugin-manifest) [![codecov](https://codecov.io/gh/khalyomede/gridsome-plugin-manifest/branch/master/graph/badge.svg)](https://codecov.io/gh/khalyomede/gridsome-plugin-manifest) ![Snyk Vulnerabilities for npm package](https://img.shields.io/snyk/vulnerabilities/npm/gridsome-plugin-manifest) ![Libraries.io dependency status for latest release](https://img.shields.io/librariesio/release/npm/gridsome-plugin-manifest) [![Maintainability](https://api.codeclimate.com/v1/badges/493c9113d81d8444ad82/maintainability)](https://codeclimate.com/github/khalyomede/gridsome-plugin-manifest/maintainability)

## Summary

- [About](#about)
- [Features](#features)
- [Requirements](#requirements)
- [Installation](#installation)
- [Usage](#usage)
- [Changelog](CHANGELOG.md)
- [API](#api)
- [Known issues](#known-issues)
- [Run the tests](#run-the-tests)

## About

I made this plugin because I use gridsome to make my next web app, and I wanted to have a way to just generate this file for me in order to pass the Lighthouse PWA manifest.json test.

## Features

- Generates a `manifest.json` at the root of your dist folder at build time
- Automatically generates 512, 192, 144, 98, 72, and 48px wide icons from your original icon
- Adds a `<meta name="theme-color" />` and a `<link rel="manifest" />` (at build time) at the head of each of your HTML files

## Requirements

- NPM or Yarn installed on your machine

## Installation

With NPM:

```bash
npm install --save-dev gridsome-plugin-manifest
```

With Yarn:

```bash
yarn add --dev gridsome-plugin-manifest
```

## Usage

In your file `gridsome.config.js`, add the following in the `plugin` key:

```javascript
module.exports = {
  siteName: "Gridsome",
  plugins: [
    {
      use: "gridsome-plugin-manifest",
      options: {
        background_color: "#000000",
        icon_path: "./src/assets/img/icon.png",
        name: "My app name",
        short_name: "App",
        theme_color: "#FFFFFF",
        lang: "en",
      },
    },
  ],
};
```

Build your project:

- With NPM: `npm run build`
- With Yarn: `yarn build`

To check it work, serve your application locally, then use Chrome DevTools to run a check on the PWA "manifest" test to check it has successfuly taken your manifest file into account.

![google-dev-tools-showing-manifest-passes-test](https://user-images.githubusercontent.com/15908747/71518859-2dc07280-28b5-11ea-8713-600a6bd542f2.png)

## API

- Options
  - **background_color**: `String` The background of your PWA loading screen
  - **name**: `String` The name displayed in your PWA loading screen
  - **theme_color**: `String` The color of the text in your PWA loading screen
  - **display**: `standalone` | `minimal-ui` | `fullscreen` (default: `minimal-ui`)
  - **scope**: `String` The scope of your PWA (should be an absolute URL)
  - **orientation**: `any` | `natural` | `landscape` | `landscape-primary` | `landscape-secondary` | `portrait` | `portrait-primary` | `portrait-secondary` (default: `any`)
  - **start_url**: `String` The URL where the user will begin if he/she starts your PWA
  - **file_name**: `String` The name of your manifest file (default: `manifest.json`)
  - **icon_path**: `String` The path (include the file name) where your icon is stored at
  - **dir**: `ltr` | `rtl` | `auto` The direction of the text (default: `auto`)
  - **lang**: `String` An ISO 2 lang code to determine which lang the `name` and `short_name` are displayed on.
  - **prefer_related_applications**: `Boolean` Whether to prompt to open any link with another app than this PWA or not (default: `false`).
  - **related_applications**: `Array<RelatedApplication>`: A list of the related applications (see below for the prototype of a `RelatedApplication`).

```typescript
interface RelatedApplication {
  platform: string;
  url: string;
  id?: string;
}
```

_(`?:` means "non mandatory field")._

## Known issues

- The icon **must** be square (the size that is used in the `icons` key of the `manifest.json` file is for the moment `width x width`, so for example a width of `512` will produce a `sizes` key equal to `512x512`)

## Run the tests

1. Clone the project: `git clone https://github.com/khalyomede/gridsome-plugin-manifest`
2. Install the dependencies
   - with NPM: `npm install`
   - with Yarn: `yarn install`
3. Run the tests
   - with NPM: `npm run test`
   - with Yarn: `yarn test`

You should see something like:

```
$ nyc mocha --require @babel/register test


  client
    √ should export a function

  server
    general
      √ should export a function
    defaultOptions
      √ should return default options


  3 passing (22ms)

-------------------|---------|----------|---------|---------|-------------------
File               | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
-------------------|---------|----------|---------|---------|-------------------
All files          |    17.8 |     4.88 |    9.52 |   18.92 |
 error-logger.js   |      80 |      100 |       0 |      80 | 8
 ...some.client.js |      50 |      100 |       0 |   66.67 | 2
 ...some.server.js |   13.76 |     4.88 |   10.53 |   14.56 | ...30-232,257-259
-------------------|---------|----------|---------|---------|-------------------
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.3.5] 2020-06-08

### Fixed

- Bug when sharp^0.24.0 would prevent the plugin to run because of an error about the file type (thank you both @vergissberlin for you PR #7 and @Kulcanhez for reporting the issue).

## [0.3.4] 2020-01-26

### Fixed

- Bug when upgrading the dependencies would cause Sharp to be built twice, on differents version, leading to libvips to not be usable by the plugin. Now, the version of sharp has been corrected to not produce 2 installation of Sharp.

## [0.3.3] 2020-01-12

### Fixed

- Building a Gridsome project using this plugin on Netlify will not make the build fail anymore (due to the removal of a package that was causing the issue).

### [0.3.2] 2020-01-05

### Fixed

- The error when you do not add an `icon_path` or have this option targeting a non existing file will now be more explicit, rather than being cryptic (because thrown by a dependency).

## [0.3.1] 2019-12-28

### Fixed

- Bug when the manifest file were not found (the error was visible in the browser console) in development mode. Instead, it will now be removed in development mode, but still generated (with its `<link rel="manifest" />`) at build time.

## [0.3.0] 2019-12-27

### Breaking

- You will need to add an ISO 2 `lang` value in your configuration in order to keep generating a manifest file without issues.

### Added

- Support for the keys `dir`, `lang`, `prefer_related_applications` and `related_applications` in the manifest: they are now configurable and will be outputed in the `manifest.json` file.

### Fixed

- The names of the different options in the API section of the README have been updated to reflect the changes in the version `0.2.0`.
- The manifest file will now be served with the MIME type `application/manifest+json` instead of `application/json` to respect the standard.

## [0.2.1] - 2019-12-26

### Fixed

- The documentation example showing the necessary keys to add to the `options` key of the plugin have been updated to follow one of the fix that the version `0.2.0` provided.

## [0.2.0] - 2019-12-26

### Added

- A `<meta name="theme-color" />` will be automatically added to the head of your html files (using the `theme_color` value you provided in the options in the file `gridsome.client.js`).

### Fixed

- The default `display` option in the manifest will now be `minimal-ui` (instead of `browser`) to fix a bug when Lighthouse was not recognizing the `browser` display value.
- The `manifest.json` file will be correctly formed (previously, pascal cased keys were used, now this file is generated using correct snake cased keys).

## [0.1.1] - 2019-12-25

### Fixed

- Bug when the `fileName` option was forbidden instead of being allowed.

## [0.1.0] - 2019-12-25

### Added

- First working version


gridsome.server.js
"use strict";

var __awaiter = void 0 && (void 0).__awaiter || function (thisArg, _arguments, P, generator) {
  function adopt(value) {
    return value instanceof P ? value : new P(function (resolve) {
      resolve(value);
    });
  }

  return new (P || (P = Promise))(function (resolve, reject) {
    function fulfilled(value) {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    }

    function rejected(value) {
      try {
        step(generator["throw"](value));
      } catch (e) {
        reject(e);
      }
    }

    function step(result) {
      result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
    }

    step((generator = generator.apply(thisArg, _arguments || [])).next());
  });
};

var __generator = void 0 && (void 0).__generator || function (thisArg, body) {
  var _ = {
    label: 0,
    sent: function sent() {
      if (t[0] & 1) throw t[1];
      return t[1];
    },
    trys: [],
    ops: []
  },
      f,
      y,
      t,
      g;
  return g = {
    next: verb(0),
    "throw": verb(1),
    "return": verb(2)
  }, typeof Symbol === "function" && (g[Symbol.iterator] = function () {
    return this;
  }), g;

  function verb(n) {
    return function (v) {
      return step([n, v]);
    };
  }

  function step(op) {
    if (f) throw new TypeError("Generator is already executing.");

    while (_) {
      try {
        if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
        if (y = 0, t) op = [op[0] & 2, t.value];

        switch (op[0]) {
          case 0:
          case 1:
            t = op;
            break;

          case 4:
            _.label++;
            return {
              value: op[1],
              done: false
            };

          case 5:
            _.label++;
            y = op[1];
            op = [0];
            continue;

          case 7:
            op = _.ops.pop();

            _.trys.pop();

            continue;

          default:
            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
              _ = 0;
              continue;
            }

            if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
              _.label = op[1];
              break;
            }

            if (op[0] === 6 && _.label < t[1]) {
              _.label = t[1];
              t = op;
              break;
            }

            if (t && _.label < t[2]) {
              _.label = t[2];

              _.ops.push(op);

              break;
            }

            if (t[2]) _.ops.pop();

            _.trys.pop();

            continue;
        }

        op = body.call(thisArg, _);
      } catch (e) {
        op = [6, e];
        y = 0;
      } finally {
        f = t = 0;
      }
    }

    if (op[0] & 5) throw op[1];
    return {
      value: op[0] ? op[1] : void 0,
      done: true
    };
  }
};

var fs_1 = require("fs");

var isHexcolor = require("is-hexcolor");

var mime = require("mime-types");

var mkdirp_1 = require("mkdirp");

var path_1 = require("path");

var rename = require("rename");

var sharp = require("sharp");

var GridsomePluginManifest = function () {
  function GridsomePluginManifest(api, options) {
    var _this = this;

    this._options = options;
    api.beforeBuild(function () {
      return __awaiter(_this, void 0, void 0, function () {
        var iconFileName, iconFileName512, iconFileName192, iconFileName144, iconFileName96, iconFileName72, iconFileName48, mimeType;
        return __generator(this, function (_a) {
          switch (_a.label) {
            case 0:
              console.time("gridsome-plugin-manifest");

              try {
                this._checkOptions();
              } catch (exception) {
                if (exception instanceof TypeError) {
                  console.error(GridsomePluginManifest.pluginName + ": " + exception.message);
                  console.timeEnd(GridsomePluginManifest.pluginName);
                } else {
                  throw exception;
                }

                return [2];
              }

              if (!fs_1.existsSync("./static/assets/img")) {
                mkdirp_1.sync("./static/assets/img");
              }

              iconFileName = path_1.basename(options.icon_path);
              iconFileName512 = rename(iconFileName, {
                suffix: "-512"
              }).toString();
              iconFileName192 = rename(iconFileName, {
                suffix: "-192"
              }).toString();
              iconFileName144 = rename(iconFileName, {
                suffix: "-144"
              }).toString();
              iconFileName96 = rename(iconFileName, {
                suffix: "-96"
              }).toString();
              iconFileName72 = rename(iconFileName, {
                suffix: "-72"
              }).toString();
              iconFileName48 = rename(iconFileName, {
                suffix: "-48"
              }).toString();
              return [4, Promise.all([sharp(options.icon_path).resize(512).toFile("./static/assets/img/" + iconFileName512), sharp(options.icon_path).resize(192).toFile("./static/assets/img/" + iconFileName192), sharp(options.icon_path).resize(144).toFile("./static/assets/img/" + iconFileName144), sharp(options.icon_path).resize(96).toFile("./static/assets/img/" + iconFileName96), sharp(options.icon_path).resize(72).toFile("./static/assets/img/" + iconFileName72), sharp(options.icon_path).resize(48).toFile("./static/assets/img/" + iconFileName48)])];

            case 1:
              _a.sent();

              if (!fs_1.existsSync("./static")) {
                mkdirp_1.sync("./static");
              }

              mimeType = mime.lookup(options.icon_path);
              mimeType = mimeType === false ? "image/png" : mimeType;
              options.icons = [{
                src: "/assets/img/" + iconFileName512,
                type: mimeType,
                sizes: "512x512"
              }, {
                src: "/assets/img/" + iconFileName192,
                type: mimeType,
                sizes: "192x192"
              }, {
                src: "/assets/img/" + iconFileName144,
                type: mimeType,
                sizes: "144x144"
              }, {
                src: "/assets/img/" + iconFileName96,
                type: mimeType,
                sizes: "96x96"
              }, {
                src: "/assets/img/" + iconFileName72,
                type: mimeType,
                sizes: "72x72"
              }, {
                src: "/assets/img/" + iconFileName48,
                type: mimeType,
                sizes: "48x48"
              }];
              fs_1.writeFileSync("./static/" + options.file_name, JSON.stringify(options, undefined, 4));
              console.timeEnd("gridsome-plugin-manifest");
              return [2];
          }
        });
      });
    });
  }

  GridsomePluginManifest.defaultOptions = function () {
    return {
      display: "minimal-ui",
      file_name: "manifest.json",
      orientation: "any",
      scope: "/",
      start_url: "/",
      dir: "auto",
      prefer_related_applications: false,
      related_applications: []
    };
  };

  GridsomePluginManifest.prototype._checkOptions = function () {
    this._checkBackgroundColorOption();

    this._checkDisplayOption();

    this._checkIconPathOption();

    this._checkNameOption();

    this._checkFileNameOption();

    this._checkOrientationOption();

    this._checkScopeOption();

    this._checkShortNameOption();

    this._checkStartUrlOption();

    this._checkThemeColorOption();

    this._checkDirOption();

    this._checkLangOption();

    this._checkPreferRelatedApplicationsOption();

    this._checkRelatedApplicationsOption();
  };

  GridsomePluginManifest.prototype._throwIfOptionMissing = function (optionName) {
    if (!(optionName in this._options)) {
      throw new TypeError("\"" + optionName + "\" must be present");
    }
  };

  GridsomePluginManifest.prototype._throwIfOptionNotString = function (optionName) {
    if (typeof this._options[optionName] !== "string") {
      throw new TypeError("\"" + optionName + "\" must be a string");
    }
  };

  GridsomePluginManifest.prototype._throwIfOptionNotFilledString = function (optionName) {
    if (this._options[optionName].length === 0) {
      throw new TypeError("\"" + optionName + "\" must be filled");
    }
  };

  GridsomePluginManifest.prototype._throwIfOptionNotOneOf = function (optionName, allowedValues) {
    if (!allowedValues.includes(this._options[optionName])) {
      throw new TypeError("\"" + optionName + "\" must be one of [" + allowedValues.toString() + "]");
    }
  };

  GridsomePluginManifest.prototype._throwIfOptionNotAbsolutePath = function (optionName) {
    if (!path_1.isAbsolute(this._options[optionName])) {
      throw new TypeError("\"" + optionName + "\" must be an absolute path");
    }
  };

  GridsomePluginManifest.prototype._throwIfOptionNotHexColor = function (optionName) {
    if (!isHexcolor(this._options[optionName])) {
      throw new TypeError("\"" + optionName + "\" must be a valid hexadecimal color");
    }
  };

  GridsomePluginManifest.prototype._checkBackgroundColorOption = function () {
    var optionName = "background_color";

    this._throwIfOptionMissing(optionName);

    this._throwIfOptionNotString(optionName);

    this._throwIfOptionNotFilledString(optionName);

    this._throwIfOptionNotHexColor(optionName);
  };

  GridsomePluginManifest.prototype._checkDisplayOption = function () {
    var optionName = "display";

    this._throwIfOptionMissing(optionName);

    this._throwIfOptionNotString(optionName);

    this._throwIfOptionNotFilledString(optionName);

    this._throwIfOptionNotOneOf(optionName, ["standalone", "minimal-ui", "fullscreen"]);
  };

  GridsomePluginManifest.prototype._checkIconPathOption = function () {
    var optionName = "icon_path";

    this._throwIfOptionMissing(optionName);

    this._throwIfOptionNotString(optionName);

    this._throwIfOptionNotFilledString(optionName);

    if (this._options.icon_path !== undefined && !fs_1.existsSync(this._options.icon_path)) {
      throw new TypeError("\"" + optionName + "\" must target an existing file");
    }
  };

  GridsomePluginManifest.prototype._checkNameOption = function () {
    var optionName = "name";

    this._throwIfOptionMissing(optionName);

    this._throwIfOptionNotString(optionName);

    this._throwIfOptionNotFilledString(optionName);
  };

  GridsomePluginManifest.prototype._checkFileNameOption = function () {
    var optionName = "file_name";

    this._throwIfOptionMissing(optionName);

    this._throwIfOptionNotString(optionName);

    this._throwIfOptionNotFilledString(optionName);
  };

  GridsomePluginManifest.prototype._checkOrientationOption = function () {
    var optionName = "orientation";

    this._throwIfOptionMissing(optionName);

    this._throwIfOptionNotString(optionName);

    this._throwIfOptionNotFilledString(optionName);

    this._throwIfOptionNotOneOf(optionName, ["any", "natural", "landscape", "landscape - primary", "landscape - secondary", "portrait", "portrait - primary", "portrait - secondary"]);
  };

  GridsomePluginManifest.prototype._checkScopeOption = function () {
    var optionName = "scope";

    this._throwIfOptionMissing(optionName);

    this._throwIfOptionNotString(optionName);

    this._throwIfOptionNotFilledString(optionName);

    this._throwIfOptionNotAbsolutePath(optionName);
  };

  GridsomePluginManifest.prototype._checkShortNameOption = function () {
    var optionName = "short_name";

    this._throwIfOptionMissing(optionName);

    this._throwIfOptionNotString(optionName);

    this._throwIfOptionNotFilledString(optionName);
  };

  GridsomePluginManifest.prototype._checkStartUrlOption = function () {
    var optionName = "start_url";

    this._throwIfOptionMissing(optionName);

    this._throwIfOptionNotString(optionName);

    this._throwIfOptionNotFilledString(optionName);

    this._throwIfOptionNotAbsolutePath(optionName);
  };

  GridsomePluginManifest.prototype._checkThemeColorOption = function () {
    var optionName = "theme_color";

    this._throwIfOptionMissing(optionName);

    this._throwIfOptionNotString(optionName);

    this._throwIfOptionNotFilledString(optionName);

    this._throwIfOptionNotHexColor(optionName);
  };

  GridsomePluginManifest.prototype._checkDirOption = function () {
    var optionName = "dir";

    this._throwIfOptionMissing(optionName);

    this._throwIfOptionNotString(optionName);

    this._throwIfOptionNotFilledString(optionName);

    this._throwIfOptionNotOneOf(optionName, ["ltr", "rtl", "auto"]);
  };

  GridsomePluginManifest.prototype._checkLangOption = function () {
    var optionName = "lang";

    this._throwIfOptionMissing(optionName);

    this._throwIfOptionNotString(optionName);

    this._throwIfOptionNotFilledString(optionName);
  };

  GridsomePluginManifest.prototype._checkPreferRelatedApplicationsOption = function () {
    var optionName = "prefer_related_applications";

    this._throwIfOptionMissing(optionName);

    if (typeof this._options.prefer_related_applications !== "boolean") {
      throw new TypeError("\"" + optionName + "\" must be a boolean");
    }
  };

  GridsomePluginManifest.prototype._checkRelatedApplicationsOption = function () {
    var optionName = "related_applications";

    this._throwIfOptionMissing(optionName);

    if (!Array.isArray(this._options.related_applications)) {
      throw new TypeError("\"" + optionName + "\" must be an array");
    }

    var numberOfRelatedApplications = this._options.related_applications.length;

    for (var index = 0; index < numberOfRelatedApplications; index++) {
      var relatedApplication = this._options.related_applications[index];

      if (!(relatedApplication instanceof Object)) {
        throw new TypeError("\"" + optionName + "[" + index + "]\" must be an object");
      }

      if (!("platform" in relatedApplication)) {
        throw new TypeError("\"" + optionName + "[" + index + "].platform\" must be present");
      }

      if (!("url" in relatedApplication)) {
        throw new TypeError("\"" + optionName + "[" + index + "].url\" must be present");
      }

      if (typeof relatedApplication.platform !== "string") {
        throw new TypeError("\"" + optionName + "[" + index + "].platform\" must be a string");
      }

      if (typeof relatedApplication.url !== "string") {
        throw new TypeError("\"" + optionName + "[" + index + "].url\" must be a string");
      }

      if ("id" in relatedApplication && typeof relatedApplication.id !== "string") {
        throw new TypeError("\"" + optionName + "[" + index + "].id\" must be a string");
      }

      if (relatedApplication.platform.length === 0) {
        throw new TypeError("\"" + optionName + "[" + index + "].platform\" must be filled");
      }

      if (relatedApplication.url.length === 0) {
        throw new TypeError("\"" + optionName + "[" + index + "].url\" must be filled");
      }

      if ("id" in relatedApplication && typeof relatedApplication.id === "string" && relatedApplication.id.length === 0) {
        throw new TypeError("\"" + optionName + "[" + index + "].id\" must be filled");
      }
    }
  };

  GridsomePluginManifest.pluginName = "gridsome-plugin-manifest";
  return GridsomePluginManifest;
}();

module.exports = GridsomePluginManifest;